test/jdk/java/net/httpclient/EmptyAuthenticate.java

```diff
@@ -24,55 +24,118 @@
+import java.util.stream.Stream;
 
-public class EmptyAuthenticate {
-
-    public static void main(String[] args) throws IOException, URISyntaxException, InterruptedException {
-        int port = 0;
-
-        //start server:
-        HttpServer server = HttpServer.create(new InetSocketAddress(port), 0);
-        port = server.getAddress().getPort();
-        server.createContext("/", exchange -> {
-            String response = "test body";
-            //this empty header will make the HttpClient throw NPE
-            exchange.getResponseHeaders().add("www-authenticate", "");
-            exchange.sendResponseHeaders(401, response.length());
-            OutputStream os = exchange.getResponseBody();
-            os.write(response.getBytes());
-            os.close();
-        });
-        server.start();
+import static java.net.http.HttpClient.Builder.NO_PROXY;
+import static org.junit.jupiter.api.Assertions.assertEquals;
 
-        HttpResponse<String> response = null;
-        //run client:
+class EmptyAuthenticate {
+
+    private static final SSLContext SSL_CONTEXT = createSslContext();
+
+    private static final String WWW_AUTH_HEADER_NAME = "WWW-Authenticate";
+
+    private static SSLContext createSslContext() {
         try {
-            HttpClient httpClient = HttpClient.newHttpClient();
-            HttpRequest request = HttpRequest.newBuilder(new URI("http://localhost:" + port + "/")).GET().build();
-            //this line will throw NPE (wrapped by IOException) when parsing empty www-authenticate response header in AuthenticationFilter:
-            response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
-            boolean ok = !response.headers().firstValue("WWW-Authenticate").isEmpty();
-            if (!ok) {
-                throw new RuntimeException("WWW-Authenicate missing");
-            }
-        } catch (IOException e) {
-            e.printStackTrace();
-            throw new RuntimeException("Test failed");
+            return new SimpleSSLContext().get();
+        } catch (IOException exception) {
+            throw new RuntimeException(exception);
+        }
+    }
+
+    @ParameterizedTest
+    @MethodSource("args")
+    void test(Version version, boolean secure) throws Exception {
+        HttpTestServer server = createStartedServerRespondingWithEmptyWwwAuthHeader(version, secure);
+        try (HttpClient client = createClient(version, secure)) {
+            HttpRequest request = createRequest(server, secure);
+            HttpResponse<Void> response = client.send(request, HttpResponse.BodyHandlers.discarding());
+            HttpHeaders responseHeaders = response.headers();
+            assertEquals(
+                    "",
+                    responseHeaders.firstValue(WWW_AUTH_HEADER_NAME).orElse(null),
+                    () -> "was expecting empty `%s` header in: %s".formatted(
+                            WWW_AUTH_HEADER_NAME, responseHeaders.map()));
```

Comment:
Hello Volkan, this appears to be the opposite of what the original test case was expecting. Before this change, we had:
```
boolean ok = !response.headers().firstValue("WWW-Authenticate").isEmpty();
if (!ok) {
    throw new RuntimeException("WWW-Authenticate missing");
}
```
So the test was expecting the response header value to be non-empty unlike this updated version. Did I misread the code?

Reply:
I'm guessing that this test passes for you locally. If so, then we might need some additional investigation here as to why it passes (given the exact opposite assertion of what the test was previously asserting).

I see that the original test was using lower case response header in the server handler and on the client side the test was reading a upper case response header. In this updated version, we use the upper case header consistently in both the handler and the client. You might want to check the original issue https://bugs.openjdk.org/browse/JDK-8263899 to see if the case of the header was playing a role in this test.

Reply:
firstValue(name) returns an Optional, so we were just testing that the Optional was not empty :-).
Now we're additionally checking that the value is the empty string.